src/main/scala/pl/project13/scala/sbt/JmhPlugin.scala

```diff
@@ -74,51 +54,32 @@ object JmhPlugin extends AutoPlugin {
         case "asm"                    => Seq("org.openjdk.jmh" % "jmh-generator-asm" % jmhVersion)    // GPLv2
         case unknown                  => throw new IllegalArgumentException(s"Unknown benchmark generator type: $unknown, please use one of the supported generators!")
       })
-    },
-
-    compile in Jmh <<= (compile in Jmh).dependsOn(generateJavaSources in Jmh, compile in Compile),
-    compile in Jmh <<= (compile in Jmh).dependsOn(compile in Compile),
-
-    run in Jmh <<= (run in Compile).dependsOn(compile in Jmh),
-    run in Compile <<= (run in Compile).dependsOn(compile in Jmh),
-
-    // Allows users to write custom runners and `runMain my.Runner -i 10 .*`.
-    // This is needed because it brings in the compiled classes onto the runs classpath
-    runMain in Compile <<= (runMain in Compile).dependsOn(compile in JmhKeys.Jmh)
+    }
   )
 
-  def generateBenchmarkJavaSources(s: TaskStreams, outputTarget: File, generatorType: String, classpath: Seq[Attributed[File]], scalaBinaryV: String): Seq[File] = {
-    require(outputTarget ne null, "outputTarget in Jmh must not be null!")
-    s.log.info(s"Generating JMH benchmark Java source files...")
+  private def generateBenchmarkSourcesAndResources(s: TaskStreams, cacheDir: File, bytecodeDir: File, sourceDir: File, resourceDir: File, generatorType: String, classpath: Seq[Attributed[File]]): (Seq[File], Seq[File]) = {
+    val inputs: Set[File] = (bytecodeDir ** "*").filter(_.isFile).get.toSet
+    val cachedGeneration = FileFunction.cached(cacheDir, FilesInfo.hash) { _ =>
```

Comment:
Awesome, thanks! I've been meaning to work on it for a while... That's exactly what we need here :)